### PR TITLE
Contributing lib/ansible/modules/network/cloudengine/ce_netconf.py module to manage HUAWEI data center CloudEngine

### PR DESCRIPTION
**Pull request by QijunPan**
add ce_netconf

##### ISSUE TYPE

 - New Module Pull Request
##### COMPONENT NAME

ce_netconf.py
##### ANSIBLE VERSION

```
ansible 2.3.0
config file = /etc/ansible/ansible.cfg
configured module search path = ['/home/ansible-2.1.1.0/lib/ansible/modules/core/network/cloudengine', '/home/ansible-2.1.1.0/lib/ansible/module_utils', '/home/ansible-2.1.1.0/lib/ansible/modules', '/home/ansible-2.1.1.0', '/home/ansible-2.1.1.0/lib/ansible/utils/']
```

##### SUMMARY

```
TASK [edit-config operation] ***************************************************
task path: /usr1/code/OpenNESS/code/current/KVM/intg/test/testcases/test-ce_netconf.yml:34
Using module file /usr/local/lib/python2.7/site-packages/ansible-2.3.0-py2.7.egg/ansible/modules/core/network/cloudengine/ce_netconf.py
<ce_6850> ESTABLISH LOCAL CONNECTION FOR USER: root
<ce_6850> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1487043783.51-104602864594431 `" && echo ansible-tmp-1487043783.51-104602864594431="` echo $HOME/.ansible/tmp/ansible-tmp-1487043783.51-104602864594431 `" ) && sleep 0'
<ce_6850> PUT /tmp/tmpMgRxvr TO /root/.ansible/tmp/ansible-tmp-1487043783.51-104602864594431/ce_netconf.py
<ce_6850> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1487043783.51-104602864594431/ /root/.ansible/tmp/ansible-tmp-1487043783.51-104602864594431/ce_netconf.py && sleep 0'
<ce_6850> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1487043783.51-104602864594431/ce_netconf.py; rm -rf "/root/.ansible/tmp/ansible-tmp-1487043783.51-104602864594431/" > /dev/null 2>&1 && sleep 0'
changed: [ce_6850] => {
    "changed": true, 
    "end_state": {
        "result": "ok"
    }, 
    "invocation": {
        "module_args": {
            "auth_pass": null, 
            "authorize": false, 
            "cfg_xml": "<config><aaa xmlns=\"http://www.huawei.com/netconf/vrp\" content-version=\"1.0\" format-version=\"1.0\"><authenticationSchemes><authenticationScheme  operation=\"create\"><authenSchemeName>default_wdz</authenSchemeName><firstAuthenMode>local</firstAuthenMode><secondAuthenMode>invalid</secondAuthenMode></authenticationScheme></authenticationSchemes></aaa></config>", 
            "host": "ce_6850", 
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", 
            "port": 12345, 
            "provider": null, 
            "rpc": "edit-config", 
            "ssh_keyfile": null, 
            "timeout": 10, 
            "transport": null, 
            "use_ssl": false, 
            "username": "rootDC", 
            "validate_certs": true
        }, 
        "module_name": "ce_netconf"
    }
}
```

### CODE REVIEW

File: lib/ansible/modules/network/cloudengine/ce_netconf.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'version': '1.0'}
+
+DOCUMENTATION = '''
+---
+module: ce_netconf
+version_added: "2.3"
+short_description: Run arbitrary netconf command on cloudengine devices.
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_netconf.py]: CloudEngine

Singular: `Run an arbitrary netconf command on CloudEngine devices.`
Plural: `Run arbitrary netconf commands on CloudEngine devices.`

File: lib/ansible/modules/network/cloudengine/ce_netconf.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+description:
+    - Sends an arbitrary netconf command to a cloudengine node and returns the results read from the device.
+author:
+    - wangdezhuang (@CloudEngine-Ansible)
+notes:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_netconf.py]: You can delete this note and instead use `required: true` under `rpc:`

File: lib/ansible/modules/network/cloudengine/ce_netconf.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.ce import get_nc_config, set_nc_config
+from ansible.module_utils.ce import execute_nc_action, ce_argument_spec, execute_nc_cli
+
+
+def main():
+    """ main """
+
+    argument_spec = dict(
+        rpc=dict(choices=['get', 'edit-config',
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_netconf.py:130]: required=True?

File: lib/ansible/modules/network/cloudengine/ce_netconf.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    - Sends an arbitrary netconf command to a cloudengine node and returns the results read from the device.
+author:
+    - wangdezhuang (@CloudEngine-Ansible)
+notes:
+    - The rpc parameter is always required.
+options:
+    rpc:
+        description:
+            - The type of rpc.
+        required: false
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_netconf.py]: Should this be required: true?

File: lib/ansible/modules/network/cloudengine/ce_netconf.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        cfg_xml=dict(required=True)
+    )
+
+    argument_spec.update(ce_argument_spec)
+    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=True)
+
+    rpc = module.params['rpc']
+    cfg_xml = module.params['cfg_xml']
+
+    if not rpc or not cfg_xml:
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_netconf.py]: You shouldn't need this if you have `required=True` on both options.

File: lib/ansible/modules/network/cloudengine/ce_netconf.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+#
+
+ANSIBLE_METADATA = {'status': ['preview'],
+                    'supported_by': 'community',
+                    'version': '1.0'}
+
+DOCUMENTATION = '''
+---
+module: ce_netconf
+version_added: "2.3"
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_netconf.py]: `version_added: "2.4"`

File: lib/ansible/modules/network/cloudengine/ce_netconf.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+short_description: Run arbitrary netconf command on cloudengine devices.
+description:
+    - Sends an arbitrary netconf command to a cloudengine node and returns the results read from the device.
```
Comment (gundalow) [inline lib/ansible/modules/network/cloudengine/ce_netconf.py]: Same review comment as `short_description`

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/cloudengine/ce_netconf.py (hunk 3 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_netconf.py:130]: YES

File: lib/ansible/modules/network/cloudengine/ce_netconf.py (hunk 5 above)
Comment (QijunPan) [reply lib/ansible/modules/network/cloudengine/ce_netconf.py]: OK,thanks

**Comment by QijunPan** [comment]
These issues have been fixed, thank you very much.

**Review by gundalow — APPROVED** [review]

**Pull request merged**